api/client.go

```diff
@@ -487,6 +463,50 @@ func NewClient(c *Config) (*Client, error) {
 	return client, nil
 }
 
+// ParseAddress transforms the provided address into a url.URL and handles
+// the case of Unix domain sockets by setting the DialContext in the
+// configuration's HttpClient.Transport.
+func ParseAddress(address string, config *Config) (*url.URL, error) {
+	u, err := url.Parse(address)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.HasPrefix(address, "unix://") {
+		// When the address begins with unix://, always change the transport's
+		// DialContext (to match previous behaviour)
+		socket := strings.TrimPrefix(address, "unix://")
+		if transport, ok := config.HttpClient.Transport.(*http.Transport); ok {
+			transport.DialContext = func(context.Context, string, string) (net.Conn, error) {
+				return net.Dial("unix", socket)
+			}
+
+			// Since the address points to a unix domain socket, the scheme in the
+			// *URL would be set to `unix`. The *URL in the client is expected to
+			// be pointing to the protocol used in the application layer and not to
+			// the transport layer. Hence, setting the fields accordingly.
+			u.Scheme = "http"
+			u.Host = socket
+			u.Path = ""
+		} else {
+			return nil, fmt.Errorf("attempting to specify unix:// address with non-transport transport")
+		}
+	} else if strings.HasPrefix(config.Address, "unix://") {
+		// When the address being set does not begin with unix:// but the previous
+		// address in the Config did, change the transport's DialContext back to
+		// use the default configuration that cleanhttp uses.
+		if transport, ok := config.HttpClient.Transport.(*http.Transport); ok {
+			transport.DialContext = (&net.Dialer{
```

Comment:
Instead of hardcoding this, why not use `cleanhttp.DefaultPooledTransport().DialContext`?

Reply:
Of course that assumes that the user didn't provide a nonstandard `HttpClient` in their `Config`.  I wonder if NewClient should copy the `DialContext` from the config it's given so that `SetAddress` can access it in this scenario?

Reply:
On your first comment, yes I like that a lot better.  On the second comment, this is what I've struggled with on this PR.  The current code unconditionally mutates the DialContext when the address being set starts with `unix://`.  I tried reducing the number of code paths where the DialContext is overwritten.  Since DialContext is set to a function, I just don't know of a good way to determine if the current DialContext is adequate or needs to be changed.